Make more stuff configurable.

arnoldrobbins opened the pull request:
Amount of base memory and extension memory, vidpal installed or not, default sectors per track and default number of cylinders in a disk can now be configured from `.freebee.toml`.

arnoldrobbins committed:
Make more stuff configurable.

philpem commented on src/main.c:
```diff
@@ -317,9 +322,13 @@ int main(int argc, char *argv[])
+	// RAM sizes come from config, default 2 Meg for each kind of memory
 	int i;
-	if ((i = state_init(2048*1024, 2048*1024)) != STATE_E_OK) {
+	int base_memory = fbc_get_int("memory", "base_memory");
```
It probably doesn't make sense to allow `base_memory` to be less than 512K. That was the lowest-memory 3B1 available (actually that was the original 7300). Most machines would have had 1MB or 2MB.
I don't know how the RAM count will behave with 1K increments either... exercise for the reader? :grin:

philpem commented on src/main.c:
```diff
@@ -317,9 +322,13 @@ int main(int argc, char *argv[])
 	int i;
-	if ((i = state_init(2048*1024, 2048*1024)) != STATE_E_OK) {
+	int base_memory = fbc_get_int("memory", "base_memory");
+	int extended_memory = fbc_get_int("memory", "extended_memory");
```
Extended memory - available as either a 512K or a 2MB RAM card, or the EIA Combo Card could be configured as 512K, 1MB or 1.5MB. I've no idea what granularity the OS uses to check for RAM... might be better to stick to 512K increments here too.

philpem commented:
Nice work :)
I don't know if it's worth restricting the RAM selections to 512K increments (i.e. what would be valid for a 3B1 upgrade). What do you think?
I'm not sure how granular the 3B1 OS will allow you to be with RAM upgrades.

agentbooth commented on src/main.c:
```diff
@@ -51,14 +51,19 @@ static int load_hd()
 		return (0);
 	} else {
-		if (wd2010_init(&state.hdc_ctx, state.hdc_disc0, 0, 512, 17, 8) == WD2010_ERR_OK) {
+		if (wd2010_init(&state.hdc_ctx, state.hdc_disc0, 0, bytes_per_sector, sectors_per_track, cylinders) == WD2010_ERR_OK) {
```
The last parameter is "heads" not cylinders -- cylinders is calculated based on hd.img filesize / (heads * sectors_per_track * bytes_per_sector)

arnoldrobbins committed:
Changes after code review.

arnoldrobbins commented:
@agentbooth Good catch on the cylinders / head. I have fixed that.

@philpem I have added memory checking in 512K increments, since it makes sense to do that.  This is ready to be reviewed again and merged if it looks good.  Then i think you can close the related issue.  Thanks!

philpem commented on src/main.c:
```diff
@@ -300,6 +300,49 @@ bool HandleSDLEvents(SDL_Window *window)
+	}
+
+	if (! base_ok) {
+		fprintf(stderr, "base_memory size %dK is invalid\n",
```
Would be good to say why it's invalid, or what the valid options are.

philpem commented:
I think that if we're having errors like "Parameter X invalid", we should have them explain to the user what would be required for the parameter to be valid. Something like: "Motherboard memory size is invalid, must be a multiple of 512K"

arnoldrobbins committed:
Improve error message for bad memory sizes.

arnoldrobbins commented:
@philpem I have revised the error messages.  Thanks.

arnoldrobbins committed:
cylinders -> heads in fbconfig.c. Oops.

arnoldrobbins committed:
Small fix in fbconfig.c.

philpem commented:
Merged - thanks!
